### PR TITLE
PP-481: make prologue hooks execute on all sister moms all the time

### PR DESCRIPTION
**Pull request by tpai26**
#### Issue-ID

* **[PP-481](https://pbspro.atlassian.net/browse/PP-481)**

#### Problem description
Added new test for prologue hook and fixed an existing test

#### Checklist:

[x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
[x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
[ ] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
[ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
[ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
[x] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
[ ] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
[x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
[ ] I have attached **test logs to the Jira ticket** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

### CODE REVIEW

File: test/tests/functional/pbs_hook_execjob_prologue.py [hunk 1]
```diff
@@ -216,5 +216,26 @@ def test_prologue_hook_set_job_attr(self):
         attr = {'resources_used.file': '2gb'}
         self.server.expect(JOB, attr, id=jid, extend='x', offset=1)
         self.server.accounting_match(
-            "E;%s;.*%s.*" % jid, 'resources_used.file=2gb', regexp=True,
+            "E;" + jid + ";.*resources_used.file=2gb.*", regexp=True,
             max_attempts=10)
+
+    def test_prologue_hook_fail_action_non_mom_hook(self):
+        """
+        test that the fail action is set to anything else then none
```
Comment (minghui-liu) [inline test/tests/functional/pbs_hook_execjob_prologue.py]: Please fix spelling. I think it should read `anything else than None`.

File: test/tests/functional/pbs_hook_execjob_prologue.py [hunk 2]
```diff
@@ -216,5 +216,26 @@ def test_prologue_hook_set_job_attr(self):
...
+        and the mom hook is not one of the follwing execjob_begin,
```
Comment (minghui-liu) [inline test/tests/functional/pbs_hook_execjob_prologue.py]: Please add colon and comma to make this sentence easier to read.

File: test/tests/functional/pbs_hook_execjob_prologue.py [hunk 3]
```diff
@@ -216,5 +216,26 @@ def test_prologue_hook_set_job_attr(self):
...
+        """
+        hook_name = "prologue"
+        hook_body = ("import pbs\n"
+                     "pbs.event().accept()\n")
+        attr = {'event': 'exechost_periodic',
+                'fail_action': 'offline_vnodes'}
+        try:
+            self.server.create_import_hook(hook_name, attr, hook_body)
+        except PbsManagerError, e:
+            exp_err="qmgr obj=prologue svr=default: " \
```
Comment (minghui-liu) [inline test/tests/functional/pbs_hook_execjob_prologue.py]: Please add spaces around `=`.

File: test/tests/functional/pbs_hook_execjob_prologue.py [hunk 4]
```diff
@@ -216,5 +216,26 @@ def test_prologue_hook_set_job_attr(self):
...
+                    "Can't set hook fail_action value to 'offline_vnodes': " \
+                    "hook event must" \
+                    " contain at least one of execjob_begin" \
+                    ", exechost_startup, execjob_prologue"
+            self.assertEqual(e.msg[0],exp_err)
```
Comment (minghui-liu) [inline test/tests/functional/pbs_hook_execjob_prologue.py]: Please add a space after the comma.

File: test/tests/functional/pbs_hook_execjob_prologue.py [hunk 5]
```diff
@@ -216,5 +216,26 @@ def test_prologue_hook_set_job_attr(self):
         attr = {'resources_used.file': '2gb'}
         self.server.expect(JOB, attr, id=jid, extend='x', offset=1)
         self.server.accounting_match(
-            "E;%s;.*%s.*" % jid, 'resources_used.file=2gb', regexp=True,
+            "E;" + jid + ";.*resources_used.file=2gb.*", regexp=True,
```
Comment (bhroam) [inline test/tests/functional/pbs_hook_execjob_prologue.py]: log match will match substrings.  You don't need the .* at the end.

File: test/tests/functional/pbs_hook_execjob_prologue.py (hunk 3 above)
Comment (bhroam) [reply test/tests/functional/pbs_hook_execjob_prologue.py]: Use the error string itself, not the qmgr portion.  If you match what qmgr returns, this test will fail in API mode.

File: test/tests/functional/pbs_hook_execjob_prologue.py [hunk 6]
```diff
@@ -216,5 +216,26 @@ def test_prologue_hook_set_job_attr(self):
...
+        attr = {'event': 'exechost_periodic',
+                'fail_action': 'offline_vnodes'}
+        try:
+            self.server.create_import_hook(hook_name, attr, hook_body)
+        except PbsManagerError, e:
+            exp_err="qmgr obj=prologue svr=default: " \
+                    "Can't set hook fail_action value to 'offline_vnodes': " \
+                    "hook event must" \
+                    " contain at least one of execjob_begin" \
+                    ", exechost_startup, execjob_prologue"
```
Comment (bhroam) [inline test/tests/functional/pbs_hook_execjob_prologue.py]: New guidelines from the PTL steering committee just came out.  Error messages should be +='d together.

File: test/tests/functional/pbs_hook_execjob_prologue.py (hunk 4 above)
Comment (bhroam) [reply test/tests/functional/pbs_hook_execjob_prologue.py]: Did you run your test case through pep8?  It usually complains about things like this.

File: test/tests/functional/pbs_hook_execjob_prologue.py [hunk 7]
```diff
@@ -223,7 +223,7 @@ def test_prologue_hook_fail_action_non_mom_hook(self):
         """
         Test that when fail action is set to anything else than 'None'
         and the mom hook is not one of the following- execjob_begin,
-        exechost_startup, execjob_prologue error msg is displayed
+        exechost_startup, execjob_prologue, error msg is displayed
```
Comment (bhroam) [inline test/tests/functional/pbs_hook_execjob_prologue.py]: How about: "Test when the fail action is set to anything other than 'None' on a mom_hook, and that mom_hook is not either execjob_begin, exechost_startup, or execjob_prologue an error message is displayed"

File: test/tests/functional/pbs_hook_execjob_prologue.py [hunk 8]
```diff
@@ -233,9 +233,8 @@ def test_prologue_hook_fail_action_non_mom_hook(self):
...
-            exp_err="qmgr obj=prologue svr=default: " \
-                    "Can't set hook fail_action value to 'offline_vnodes': " \
-                    "hook event must" \
-                    " contain at least one of execjob_begin" \
-                    ", exechost_startup, execjob_prologue"
-            self.assertEqual(e.msg[0],exp_err)
+            exp_err = "Can't set hook fail_action value to 'offline_vnodes':"
+            exp_err = exp_err + " hook event must"
+            exp_err = exp_err + " contain at least one of execjob_begin"
+            exp_err = exp_err + ", exechost_startup, execjob_prologue"
```
Comment (bhroam) [inline test/tests/functional/pbs_hook_execjob_prologue.py]: I believe the guideline was to to use var += msg, not var = var + msg.

File: test/tests/functional/pbs_hook_execjob_prologue.py [hunk 9]
```diff
@@ -216,5 +216,25 @@ def test_prologue_hook_set_job_attr(self):
...
+        attr = {'event': 'exechost_periodic',
+                'fail_action': 'offline_vnodes'}
+        try:
+            self.server.create_import_hook(hook_name, attr, hook_body)
+        except PbsManagerError, e:
+            exp_err = "Can't set hook fail_action value to 'offline_vnodes':"
+            exp_err += " hook event must"
+            exp_err += " contain at least one of execjob_begin"
+            exp_err += ", exechost_startup, execjob_prologue"
+            self.assertTrue(exp_err in e.msg[0])
```
Comment (minghui-liu) [inline test/tests/functional/pbs_hook_execjob_prologue.py]: `e.msg` is a string rather than a list.

Comment (bhroam) [reply test/tests/functional/pbs_hook_execjob_prologue.py]: You can do it.  It just does a substring match.  I do this all the time.  If this is not the right way of doing a substring match, we should get guidance from the PTL steering committee.

Comment (minghui-liu) [reply test/tests/functional/pbs_hook_execjob_prologue.py]: `exp_err in e.msg` would work. But `e.msg[0]` will give you the first character of the string.

Comment (bhroam) [reply test/tests/functional/pbs_hook_execjob_prologue.py]: e.msg is a list.  It's a list of lines returned from qmgr.  e.msg[0] is the first line.  This is where the error message from PBS is returned.  For CLI mode, there's extra text added on by the qmgr command.  For the API it's just the error message itself.  This is why the substring match works.

**Comment by bhroam** [comment]
@tpai26 your branch is out of date with master.  Please rebase again.

**Comment by bhroam** [comment]
I merged in the PR.

**Pull request merged**